tests/test_routes.py

```diff
@@ -0,0 +1,633 @@
+
+        # Locate the function itself.
+        function = getattr(routes, function_name)
+
+        # Call it in the context of our MockApp which simulates the
+        # controller code.
+        with self.app.test_request_context():
+            return function(**kwargs)
+
+    def assert_request_calls(self, url, method, *args, **kwargs):
+        """Make a request to the given `url` and assert that
+        the given controller `method` was called with the
+        given `args` and `kwargs`.
+        """
+        http_method = kwargs.pop('http_method', 'GET')
+        response = self.request(url, http_method)
+        eq_(response.method, method)
+        eq_(response.method.args, args)
+        eq_(response.method.kwargs, kwargs)
+
+    def assert_authenticated_request_calls(self, url, method, *args, **kwargs):
+        """First verify that an unauthenticated request fails. Then make an
+        authenticated request to `url` and verify the results, as with
+        assert_request_calls
+        """
+        http_method = kwargs.pop('http_method', 'GET')
+        response = self.request(url, http_method)
+        eq_(401, response.status_code)
+        eq_("authenticated_patron_from_request called without authorizing",
+            response.data)
+
+        # Set a variable so that authenticated_patron_from_request
+        # will succeed, and try again.
+        self.manager.index_controller.authenticated = True
+        try:
+            kwargs['http_method'] = http_method
+            self.assert_request_calls(url, method, *args, **kwargs)
+        finally:
+            # Un-set authentication for the benefit of future
+            # assertions in this test function.
+            self.manager.index_controller.authenticated = False
+
+    def assert_supported_methods(self, url, *methods):
+        """Verify that the given HTTP `methods` are the only ones supported
+        on the given `url`.
+        """
+        # The simplest way to do this seems to be to try each of the
+        # other potential methods and verify that MethodNotAllowed is
+        # raised each time.
+        check = set(['GET', 'POST', 'PUT', 'DELETE']) - set(methods)
+
+        # Treat HEAD specially. Any controller that supports GET
+        # automatically supports HEAD. So we only assert that HEAD
+        # fails if the method supports neither GET nor HEAD.
+        if 'GET' not in methods and 'HEAD' not in methods:
+            check.add('HEAD')
+        for method in check:
+            logging.debug("MethodNotAllowed should be raised on %s", method)
+            assert_raises(MethodNotAllowed, self.request, url, method)
+            logging.debug("And it was.")
+
+
+class TestIndex(RouteTest):
+
+    CONTROLLER_NAME = "index_controller"
+
+    def test_index(self):
+        for url in '/', '':
+            self.assert_request_calls(url, self.controller)
+
+    def test_authentication_document(self):
+        url = '/authentication_document'
```

Comment:
Throughout these tests I used the actual URL templates for the routes whenever possible, so it would be easy to match the test with the original code. The only exception is when a template included `<path:identifier>`, which I converted to `an/identifier`. This lets us verify that the <path:> thing works.